numpy/core/include/numpy/ufuncobject.h

```diff
@@ -111,6 +111,24 @@ typedef int (PyUFunc_MaskedInnerLoopSelectionFunc)(
                             NpyAuxData **out_innerloopdata,
                             int *out_needs_api);
 
+typedef struct _ufunc_extension {
+        /* New in PyUFuncObject version 1 and above */
+
+        /*
+         * for each core_num_dim_ix distinct dimension names,
+         * the possible "frozen" size (-1 if not frozen).
+         */
+        npy_intp *core_dim_sizes;
+
+        /*
+         * for each distinct core dimension, a set of flags OR'd together
+         * e.g., UFUNC_CORE_DIM_CAN_IGNORE if signature has ?
+         *       UFUNC_CORE_DIM_SIZE_UNSET for non-frozen dimensions.
```

Comment:
Why not just `UFUNC_CORE_DIM_SIZE_FROZEN` to mean the inverse?

Reply:
@mhvk any opinions? I could not find a place where we explicitly use the flag value except in test_signature

Reply:
I had is as `FROZEN` initially, but changed it because that meant that `0` was something for which the code does not have to do anything - i.e., all flags mean that some special action has to be taken (in this case to get the dimension from the inputs). Of course, here it doesn't matter all that much size really the way this is signaled is by making the preset size negative. Still, personally I like the logic. (But I think it became a bit more logical in the follow-up PR, which allowed inner dimensions to be broadcast.)